### PR TITLE
add XCTVapor module

### PR DESCRIPTION
**Pull request by tanner0101**
Adds a new `XCTVapor` testing module. This module revolves around an `XCTApplication` class that wraps a Vapor `Application` and provides utilities for testing.

```swift
import App
import XCTVapor

XCTVapor.app = { try App.app(.testing) }

class MyApplicationTests: XCTVaporTests {
    func testExample() throws {
        print(self.app) // Application
        try self.app.xctest()
            .test(.GET, to: "/") { res in
                res.assertStatus(is: .ok)
                    .assertBody(contains: "It works")
            }
            .test(.GET, to: "/hello") { res in
                res.assertStatus(is: .ok)
                    .assertBody(contains: "Hello, world")
            }
            .test(.GET, to: "/foo") { res in
                res.assertStatus(is: .notFound)
                    .assertBody(contains: "Not Found")
        }
    }
}
```

There are just some basic assertions here at the moment, as a proof of concept. 

Some additional ideas that I would like to implement:

-  Live option, where tests will be made against a running server instead of in-memory responder

```swift
try self.app.xctest(live: true)
    .test(.GET, to: "/") { ... }
```

### CODE REVIEW

**Comment by MrLotU** [comment]
Name ideas: 
 + `TestableVapor`
 + `VaporTestHelpers`

**Comment by calebkleveter** [comment]
Name ideas:

- `VaporAssert`
- `VaporAssess`

File: Sources/XCTVapor/XCTHTTPResponse.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    public func assertStatus(is status: HTTPStatus, file: StaticString = #file, line: UInt = #line) -> XCTHTTPResponse {
+        XCTAssertEqual(self.response.status, status, file: file, line: line)
+        return self
+    }
+    
+    @discardableResult
+    public func assertBody(contains string: String, file: StaticString = #file, line: UInt = #line) -> XCTHTTPResponse {
+        let bodyString = self.response.body.description
+        if !bodyString.contains(string) {
+            XCTFail("body contains check: (\(string.debugDescription)) does not appear in (\(bodyString.debugDescription))", file: file, line: line)
```
Comment (MrLotU) [inline Sources/XCTVapor/XCTHTTPResponse.swift]: Bit of confusing text here I think
`body contains check:`, but it does NOT contain the check 🤔

Comment (tanner0101) [reply Sources/XCTVapor/XCTHTTPResponse.swift]: ah yeah, that does read badly. it's meant to be like "the check of `bodyContains`". any ideas on how to improve?

File: Sources/XCTVapor/XCTApplication.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    @discardableResult
+    public func assert(
+        _ method: HTTPMethod,
+        to string: String,
+        file: StaticString = #file,
+        line: UInt = #line,
+        closure: (XCTHTTPResponse) throws -> ()
+    ) throws -> XCTApplication {
+        let res = try self.c.make(Responder.self).respond(
+            to: .init(method: .GET, urlString: string),
```
[the remark this answers is not in the document]
Comment (tanner0101) [reply Sources/XCTVapor/XCTApplication.swift]: d'oh, thanks

File: Sources/XCTVapor/XCTApplication.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            using: .init(channel: EmbeddedChannel())
+            ).wait()
+        try closure(.init(response: res))
```
[the remark this answers is not in the document]
Comment (tanner0101) [reply Sources/XCTVapor/XCTApplication.swift]: I don't think so. We need to specifically initialize `XCTHTTPResponse` here which is a reference type wrapper around `HTTPResponse` that has all the assert methods.

[the remark this answers is not in the document]
Comment (MrLotU) [reply Sources/XCTVapor/XCTApplication.swift]: Yeah, was afraid of that, not a big deal though :)

File: Sources/XCTVapor/XCTHTTPResponse.swift (hunk 1 above)
Comment (MrLotU) [reply Sources/XCTVapor/XCTHTTPResponse.swift]: `body contains check failed: ...` would work I think

Comment (MrLotU) [reply Sources/XCTVapor/XCTHTTPResponse.swift]: That way it clearly indicates the failure

Comment (tanner0101) [reply Sources/XCTVapor/XCTHTTPResponse.swift]: keep in mind this string will be prefixed by: `"test failed - "` from `XCTFail`

Comment (MrLotU) [reply Sources/XCTVapor/XCTHTTPResponse.swift]: Ah, in that case I don't think it has to be changed. It'd be readable and clear enough :)

**Comment by 0xTim** [comment]
I like the idea! Having some additional helpers to send and receive responses (and decode them) would be great. (Something like https://github.com/raywenderlich/vapor-til/blob/master/Tests/AppTests/Application%2BTestable.swift#L58-L137)

My main concern at the moment is the lack of injection points for tests. These need to be opened up for it to be useful. For instance if I want to use a `FakeClient` in my tests, how do I go about setting the `Application` up with that? The same applies for in-memory repositories, `FakeRenderer`s etc that you might need. I haven't looked into how dependency injection will work in Vapor 4, but it should be relatively simple if it's dependency injection or service locator etc

**Comment by penny-coin** [comment]
Hey @tanner0101, you just merged a pull request, have a coin! 

You now have 1105 coins.

**Comment by valeriomazzeo** [comment]
@tanner0101 why `XCTVapor.app = { try App.app(.testing) }` has to be static? Being able to specify the way the app is created per each `XCTVaporTests` case is way more useful.

**Comment by tanner0101** [comment]
You could do that by overriding `XCTVaporTests.setUp()` right?

**Pull request merged**